api/envoy/config/cluster/v3/filter.proto

```diff
@@ -23,9 +25,16 @@ message Filter {
   // The name of the filter configuration.
   string name = 1 [(validate.rules).string = {min_len: 1}];
 
-  // Filter specific configuration which depends on the filter being
-  // instantiated. See the supported filters for further documentation.
-  // Note that Envoy's :ref:`downstream network
-  // filters <config_network_filters>` are not valid upstream filters.
-  google.protobuf.Any typed_config = 2;
+  oneof config_type {
+    // Filter specific configuration which depends on the filter being
+    // instantiated. See the supported filters for further documentation.
+    // Note that Envoy's :ref:`downstream network
+    // filters <config_network_filters>` are not valid upstream filters.
+    google.protobuf.Any typed_config = 2;
+
+    // Configuration source specifier for an extension configuration discovery
+    // service. In case of a failure and without the default configuration, the
+    // listener closes the connections.
+    core.v3.ExtensionConfigSource config_discovery = 3;
```

Comment:
As a stable API, we can't upgrade to `oneof`, as per https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md#backwards-compatibility. Instead just note this in comments and/or enforce in code.

Reply:
Ack, done